scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -0,0 +1,80 @@
+package com.twitter.scalding.beam_backend
+
+import com.twitter.scalding.Config
+import com.twitter.scalding.typed.TypedSource
+import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.transforms.DoFn.ProcessElement
+import org.apache.beam.sdk.transforms._
+import org.apache.beam.sdk.values.PCollection
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext
+
+sealed abstract class BeamOp[+A] {
+  import BeamOp.TransformBeamOp
+  def run(pipeline: Pipeline)(implicit ec: ExecutionContext): PCollection[_ <: A]
+
+  def map[C >: A, B](f: C => B): BeamOp[B] = {
+    this.parDo(new MapFn(f))
+  }
+  def parDo[C >: A, B](f: DoFn[C, B]): BeamOp[B] = {
+    this.applyPTransform(ParDo.of(f).asInstanceOf[PTransform[PCollection[C], PCollection[B]]])
+  }
+  def applyPTransform[C >: A, B](f: PTransform[PCollection[C], PCollection[B]]): BeamOp[B] = {
+    TransformBeamOp(this, f)
+  }
+  def filter(f: A => Boolean): BeamOp[A] = {
+    val fn = new ProcessFn(f.asInstanceOf[A => java.lang.Boolean])
+    this.applyPTransform(Filter.by[A, ProcessFunction[A, java.lang.Boolean]](fn))
+  }
+  def flatMap[C >: A, B](f: C => TraversableOnce[B]): BeamOp[B] = {
+    this.parDo(new FlatMapFn(f))
+  }
+}
+
+object BeamOp extends Serializable {
+
+  final case class Source[A](
+    conf: Config,
+    original: TypedSource[A],
+    input: Option[BeamSource[A]]) extends BeamOp[A] {
+    def run(pipeline: Pipeline)(implicit ec: ExecutionContext): PCollection[_ <: A] =
+      input match {
+        case None => throw new IllegalArgumentException(s"source $original was not connected to a beam source")
+        case Some(src) => src.read(pipeline, conf)
+      }
+  }
+
+  final case class FromIterable[A](iterable: Iterable[A]) extends BeamOp[A] {
+    override def run(pipeline: Pipeline)
+      (implicit ec: ExecutionContext): PCollection[_ <: A] = {
+      pipeline.apply(Create.of(iterable.asJava).withCoder(new KryoCoder[A]))
+    }
+  }
+
+  final case class TransformBeamOp[A, B](source: BeamOp[A], f: PTransform[PCollection[A], PCollection[B]]) extends BeamOp[B] {
+    def run(pipeline: Pipeline)(implicit ec: ExecutionContext): PCollection[B] = {
+      source.run(pipeline).asInstanceOf[PCollection[A]].apply(f)
+        .setCoder(new KryoCoder[B])
+    }
+  }
+
+
+}
+class ProcessFn[A, B](f: A => B) extends ProcessFunction[A, B] {
```

Comment:
also probably better to make a `case class` here just to have some chance of plumbing through the equality and potentially getting a clearer toString if `f` is one of our other internal functions.

How is this different from `MapFn`? Why do we need both? Can you make a comment?

Reply:
We need ProcessFn just because Beam Filter needs it as a lambda function that extends an `apply` method, we could also use ParDo - DoFn with MapFn but then it will be just duplicated code from Beam Filter